Extensions/AuthPropertiesTokenCacheHelper.cs

```diff
@@ -1,12 +1,57 @@
-﻿using System;
-using System.Security.Claims;
-using Microsoft.AspNetCore.Authentication;
+﻿using Microsoft.AspNetCore.Authentication.AzureAD.UI;
 using Microsoft.AspNetCore.Authentication.Cookies;
 using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Identity.Client;
+using System;
+using System.Security.Claims;
 
 namespace Microsoft.AspNetCore.Authentication
 {
+    /// <summary>
+    /// Extension class enabling adding the CookieBasedTokenCache implentation service
+    /// </summary>
+    public static class CookieBasedTokenCacheExtension
+    {
+        /// <summary>
+        /// Add the token acquisition service.
+        /// </summary>
+        /// <param name="services">Service collection</param>
+        /// <returns>the service collection</returns>
+        public static IServiceCollection AddCookieBasedTokenCache(this IServiceCollection services)
+        {
+            // Token acquisition service
+            services.AddSingleton<ITokenCacheProvider, CookieTokenCacheProvider>();
+            return services;
+        }
+    }
+
+    /// <summary>
+    /// Provides an implementation of <see cref="ITokenCacheProvider"/> for a cookie based token cache implementation
+    /// </summary>
+    public class CookieTokenCacheProvider : ITokenCacheProvider
+    {
+        /// <summary>
+        /// Get an MSAL.NET Token cache from the HttpContext, and possibly the AuthenticationProperties and Cookies sign-in scheme
+        /// </summary>
+        /// <param name="httpContext">HttpContext</param>
+        /// <param name="authenticationProperties">Authentication properties</param>
+        /// <param name="signInScheme">Sign-in scheme</param>
+        /// <returns>A token cache to use in the application</returns>
+
+        public TokenCache GetCache(HttpContext httpContext, AuthenticationProperties authenticationProperties, string signInScheme)
+        {
+            if (authenticationProperties!=null)
+            {
+                return AuthPropertiesTokenCacheHelper.ForCodeRedemption(authenticationProperties);
+            }
+            else
+            {
+                return AuthPropertiesTokenCacheHelper.ForApiCalls(httpContext, signInScheme?? AzureADDefaults.CookieScheme);
```

Comment:
I am not sure this is the best place to have this dichotomy. It looks like you are able to get AuthProperties from a SignInScheme, which suggests an adapter pattern. That being said, I do not know if you can invoke the adapt code sooner than on BeforeWrite notification arg.